File: src/main/com/dune/ide/settings/DuneColorSettingsPage.java
package com.dune.ide.settings;

import com.dune.DuneLanguage;
import com.dune.icons.DuneIcons;
import com.dune.ide.highlight.DuneSyntaxHighlighter;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighter;
import com.intellij.openapi.options.colors.AttributesDescriptor;
import com.intellij.openapi.options.colors.ColorDescriptor;
import com.intellij.openapi.options.colors.ColorSettingsPage;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.HashMap;
import java.util.Map;

/**
 * Configure colors for dune
 */
public class DuneColorSettingsPage implements ColorSettingsPage {
    private static final AttributesDescriptor[] DESCRIPTORS = new AttributesDescriptor[]{
            new AttributesDescriptor("Comment", DuneSyntaxHighlighter.COMMENT_),
            new AttributesDescriptor("Stanza", DuneSyntaxHighlighter.STANZAS_),
            new AttributesDescriptor("Fields", DuneSyntaxHighlighter.FIELDS_),
            new AttributesDescriptor("Options", DuneSyntaxHighlighter.OPTIONS_),
            new AttributesDescriptor("Atoms", DuneSyntaxHighlighter.ATOM_),
            new AttributesDescriptor("Variables", DuneSyntaxHighlighter.VAR_),
            new AttributesDescriptor("Parenthesis", DuneSyntaxHighlighter.PARENS_),
    };

    private static final Map<String, TextAttributesKey> additionalTags = new HashMap<>();

    static {
        additionalTags.put("csField", DuneSyntaxHighlighter.FIELDS_);
        additionalTags.put("csStanza", DuneSyntaxHighlighter.STANZAS_);
        additionalTags.put("csVar", DuneSyntaxHighlighter.VAR_);
    }

    @Override public @Nullable Icon getIcon() {
        return DuneIcons.FileTypes.DUNE_FILE;
    }

    @Override public @NotNull SyntaxHighlighter getHighlighter() {
        return new DuneSyntaxHighlighter();
    }

    @Override public @NonNls @NotNull String getDemoText() {
        return "; A single line comment\n\n"
                + "#| Block comments #| can be \"nested\" |# |#\n\n"
                + "(<csStanza>executable</csStanza>\n"
                + "  (<csField>names</csField> (main))\n"
                + "  #; (this S-expression\n"
                + "         (has been commented out)\n"
                + "       )\n"
                + "  (<csField>libraries</csField> (hello_world)))\n\n"
                + "(<csStanza>install</csStanza>\n"
                + "  (<csField>section</csField> bin)\n"
                + "  (<csField>files</csField> ((main.exe as hello_world))))\n\n"
                + "(<csStanza>rule</csStanza>\n"
                + "  (<csField>targets</csField> (config.full)\n"
                + "  (<csField>deps</csField>    (config_common.ml config))\n"
                + "  (<csField>action</csField>  (run <csVar>%{OCAML}</csVar> <csVar>%{path:real_configure.ml}</csVar>)))";
    }

    @Override public @Nullable Map<String, TextAttributesKey> getAdditionalHighlightingTagToDescriptorMap() {
        return additionalTags;
    }

    @Override public AttributesDescriptor @NotNull [] getAttributeDescriptors() {
        return DESCRIPTORS;
    }

    @Override public ColorDescriptor @NotNull [] getColorDescriptors() {
        return ColorDescriptor.EMPTY_ARRAY;
    }

    @Override public @NotNull String getDisplayName() {
        return DuneLanguage.NAME;
    }
}

File: src/main/kotlin/com/dune/ide/settings/DuneColorSettingsPage.kt
package com.dune.ide.settings

import com.dune.icons.DuneIcons
import com.dune.ide.colors.DuneColor
import com.dune.ide.highlight.DuneSyntaxHighlighter
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighter
import com.intellij.openapi.options.colors.AttributesDescriptor
import com.intellij.openapi.options.colors.ColorDescriptor
import com.intellij.openapi.options.colors.ColorSettingsPage
import javax.swing.Icon

class DuneColorSettingsPage : ColorSettingsPage {
    override fun getDisplayName(): String = "Dune"
    override fun getIcon(): Icon = DuneIcons.FileTypes.DUNE_FILE
    override fun getAttributeDescriptors() = Constants.ATTRS
    override fun getColorDescriptors(): Array<ColorDescriptor> = ColorDescriptor.EMPTY_ARRAY
    override fun getHighlighter(): SyntaxHighlighter = DuneSyntaxHighlighter()
    override fun getAdditionalHighlightingTagToDescriptorMap() = Constants.ANNOTATOR_TAGS
    override fun getDemoText() = Constants.DEMO_TEXT

    internal object Constants {
        val ATTRS: Array<AttributesDescriptor> = DuneColor.values().map{ it.attributesDescriptor }.toTypedArray()
        val ANNOTATOR_TAGS: Map<String, TextAttributesKey> = DuneColor.values().associateBy({ it.name }, { it.textAttributesKey })
        val DEMO_TEXT: String by lazy {
            """; Simple Dune File
(library
 (name my_lib)
 (modules my_module))

(executable
 (name "my_executable")
 (libraries my_lib)
 (modules main))"""
        }
    }

}